app/src/main/java/io/appium/uiautomator2/utils/AXWindowHelpers.java

```diff
@@ -85,13 +87,20 @@ public static void refreshRootAXNode() {
      */
     public static AccessibilityNodeInfo[] getWindowRoots() {
         List<AccessibilityNodeInfo> ret = new ArrayList<>();
+
         /*
-         * TODO: MULTI_WINDOW is disabled, UIAutomatorViewer captures active window properties and
-         * end users always relay on UIAutomatorViewer while writing tests.
+         * TODO: MULTI_WINDOW is disabled by default, UIAutomatorViewer captures active window
+         * properties and end users always relay on UIAutomatorViewer while writing tests.
          * If we enable MULTI_WINDOW it effects end users.
          * https://code.google.com/p/android/issues/detail?id=207569
          */
-        if (CustomUiDevice.getInstance().getApiLevelActual() >= Build.VERSION_CODES.LOLLIPOP && MULTI_WINDOW) {
+
+        boolean multiWindowsEnabled = AppiumUIA2Driver
+                .getInstance()
+                .getSessionOrThrow()
+                .getCapability(ENABLE_MULTI_WINDOWS.toString(), false);
+
+        if (CustomUiDevice.getInstance().getApiLevelActual() >= Build.VERSION_CODES.LOLLIPOP && multiWindowsEnabled) {
             // Support multi-window searches for API level 21 and up
             for (AccessibilityWindowInfo window : CustomUiDevice.getInstance().getInstrumentation()
                     .getUiAutomation().getWindows()) {
```

Comment:
shouldn't we try to refresh the accessibility if window.getRoot == null?

Reply:
I guess yes. I'm not familiar with that logic, I'll have a look tomorrow

Reply:
nevermind - clearing the accessibility cache here might invalidate previously retrieved roots. Lets keep it as is

Reply:
We could try

```java
        if (CustomUiDevice.getInstance().getApiLevelActual() >= Build.VERSION_CODES.LOLLIPOP && MULTI_WINDOW) {
            boolean wasAccessibilityRefreshed = false;
            do {
                boolean shouldRefreshAccessibility = false;
                // Support multi-window searches for API level 21 and up
                for (AccessibilityWindowInfo window : CustomUiDevice.getInstance().getInstrumentation().getUiAutomation().getWindows()) {
                    AccessibilityNodeInfo root = window.getRoot();
                    if (root == null && !wasAccessibilityRefreshed) {
                        shouldRefreshAccessibility = true;
                        break;
                    }
                    if (root == null) {
                        Logger.debug(String.format("Skipping null root node for window: %s", window.toString()));
                        continue;
                    }
                    ret.add(root);
                }

                if (!shouldRefreshAccessibility) {
                    break;
                }

                ret.clear();
                Device.waitForIdle();
                clearAccessibilityCache();
                wasAccessibilityRefreshed = true;
            } while (true);
        } else {
            ret.add(currentActiveWindowRoot());
        }
```

Reply:
Lets keep it as is for now, we might change the implementation later if we observe any problems with it